Reduce boot delay in loader.conf. Add custom loader.conf

bretton opened the pull request:
Tested with an RC build, works fine. Quicker boot process.

Do we want this?

bretton committed:
Reduce boot delay in loader.conf. Add custom loader.conf

grembo commented on build.sh:
```diff
@@ -32,6 +32,7 @@ fi
 # get command line flags
 while getopts i:f:u: flag
 do
+    # shellcheck disable=SC2220 # XXX: should this be a to-do, cater for wildcard?
```
We definitely should have the standard unknown option/show usage handler here.

bretton replied:
this is resolved in latest commit

grembo commented on build.sh:
```diff
@@ -70,6 +71,7 @@ MYCUSTOMDIR="${BASEDIR}/customfiles"
 cd "${BASEDIR}" || exit
 
 # check remote settings
+# shellcheck disable=SC1090 # XXX: to-do? use '. file' format instead of 'source file'?
```
As this is written in bash, `source` is ok.

I would suggest to use this line instead of disable=SC1090 and place it directly over line 76 (source...)

```
# shellcheck source=/dev/null.
```

bretton replied:
this is added

grembo commented on build.sh:
```diff
@@ -168,6 +170,7 @@ fi
 make iso BASE="${MYBASE}" RELEASE="${MYRELEASE}" ARCH="${MYARCH}" ROOTPW_HASH="*"
 
 # scp to distribution site
+# shellcheck disable=SC2154 # this relates to above, can ignore?
```
I would suggest to remove the `shellcheck disable line` here and solve the underlying issues:

1. MYREMOTECONFIG should be set to `settings.sh` instead of settings.cfg to make it clear that it's sourcing a script (not strictly needed, but nicer for the user as it makes things clear)
2.  -u be turned from an option with arguments to a flag (so one would say `./build.sh -u` to upload). Basically:
```
  u) UPLOAD="YES" ;;
  ...
   if [ "$UPLOAD" = "YES" ]; then
   ...
   fi
```
3. All things set by settings.sh should be initialized to empty or good default values(e.g., a good default value for CFG_REMOTEPORT is empty, so that what is configured for the host in .ssh/config catches on) before importing them. I would still use UPPERCASE for all settings, as they're shell variables that are pulled in - maybe prefix them with `CFG_`
4. Within the "if UPLOAD" block, required parameters should be checked.

What I would suggest is to only allow setting:
- CFG_SSH_REMOTEHOST
- CFG_SSH_REMOTEPATH (default empty)

and leave all remaining up to .ssh/config.

If you allow overriding more values (for reasons), make sure they're not used at all if they're not set (bash makes this really easy). This will make sure that entries in ~/.ssh/config are used unless explicitly overridden by the user. E.g.

```
ssh_args=()
if [ -n "$CFG_SSH_REMOTEPORT" ]; then
  ssh_args+=(-p "$CFG_SSH_REMOTEPORT")
fi
# ...more assembly of options...
scp "${ssh_args[@]}" 
```

I can help you with scripting details if you get stuck.

bretton replied:
I have a workable solution here

grembo commented:
> Tested with an RC build, works fine. Quicker boot process.
> 
> Do we want this?

Quicker boot process sounds good in general (it's already pretty good though). Not sure if there are situations where a user wants to drop to a boot loader - but if they want, they probably can also use a vanilla mfsbsd image. So I would say yes, please!

bretton committed:
adding shellcheck exclusions to fix CI issues

bretton committed:
Many fixups to script

'set -eo pipefail' caught me by surprise as grep for cd-rom was
causing script to silently exit and no indication of problem.

found a workaround with ```cmd | { grep term || :; } | cmd```

tested and works with rc1

bretton committed:
fixing spaces to tab

bretton committed:
fixing path to local ssh config

bretton committed:
fixing typo in final iso unmount

bretton commented:
successful build test. both key options work in run test

bretton committed:
Fixing up errors in user feedback

User feedback asks for -k but case statement was looking for -f

Fixed space in shellcheck comment

grembo commented on README.md:
````diff
@@ -12,19 +10,18 @@ git clone --recurse-submodules https://github.com/depenguin-me/depenguin-builder
 
 ### Set your remote host settings
-Set remote host settings
+Set remote host settings in ```~/.ssh/config```, then set remote host and path as follows:
````
Just a remark on style (not a hard request to change, as everyone has their own style):

In markdown, tripple backticks (```) are not needed for inline content, so usually using single backticks looks nicer, sobasically it's

```markdown
    This is my text with a `command` in it.
```
instead of 

```markdown
    This is my text with a ```command``` in it.
```

bretton replied:
happy to use this approach

grembo commented on README.md:
````diff
@@ -12,19 +10,18 @@ git clone --recurse-submodules https://github.com/depenguin-me/depenguin-builder
+CFG_SSH_REMOTEPATH="/path/to/www"
 ```
-The script will ```scp``` the output image file to the location specified if set to do so.
+The script will ```scp``` the output image file to the location specified if set to do so with the ```-u``` parameter.
````
Another style remark (hence not a requirement to change): I personally prefer to use indentation with four spaces instead of tripple backticks _if the snippet is short_. This makes reading easier when just looking at text at the console.

Therefore this whole section could look like below:
```markdown
Set remote host settings in `~/.ssh/config`, then set remote host
and path as follows:

    cd depenguin-builder
    cp settings.sh.sample settings.sh

and configure `settings.sh`

    cat >settings.sh <<EOF
    CFG_SSH_REMOTEHOST="your.remote.host"' 
    CFG_SSH_REMOTEPATH="/path/to/www"
    EOF

The script will `scp` the output image file to the location
specified if set to do so with the `-u` parameter.
```

bretton replied:
ditto, will use this

grembo commented on README.md:
```diff
@@ -12,19 +10,18 @@ git clone --recurse-submodules https://github.com/depenguin-me/depenguin-builder
-remotehost="your.remote.host"
-remoteport="22"
-remotepath="/path/to/www/files/"
+#!/usr/bin/env bash
```
This is actually a request to change documentation content-wise: The shebang line isn't necessary when sourcing files and it certainly looks a bit cleaner without it.

grembo commented on build.sh:
```diff
@@ -5,80 +5,109 @@
+	Usage: $(basename "${BASH_SOURCE[0]}") [-hbu] [-k /path/to/authorized_keys] 
+	
+	-h Show help
+	-b Build without uploading
```
Shouldn't this just be the default and the "-b" option disappear?

grembo commented on build.sh:
```diff
@@ -5,80 +5,109 @@
+           usage
+           exit 0
+           ;;
+        b)
```
I would simple add "UPLOAD=NO" above the `while` loop and remove "-b"

If we want to offer both, I would suggest to add a subcommand that can work independently, e.g.:

```
./build.sh build     # only build
./build.sh upload    # only upload a previous build
./build.sh build upload    # both in one go
```

But this would probably be overkill. Another option would be to move the upload functionality to a separate script, so one could run:

    ./build.sh && ./upload.sh

But the quick solution to me would be ti simply remove "-b"

grembo commented on build.sh:
```diff
@@ -93,8 +122,9 @@ if [ ! -f "${BASEDIR}/${FREEBSDISOFILE}" ]; then
 fi
 
 # mount the iso file
+# shellcheck disable=SC2086
```
This line can be deleted (disable=SC2086) as you properly escape in the mount command

grembo replied:
I really think this line can go away. Would shellcheck complain to you otherwise (my local copy didn't when I removed it)?

grembo commented on build.sh:
```diff
@@ -164,18 +195,24 @@ if [ -f "${OUTISO}" ]; then
+    make iso BASE="${MYBASE}" RELEASE="${MYRELEASE}" ARCH="${MYARCH}" ROOTPW_HASH="*"
+else
+    exit_error "UPLOAD is unset"
+fi
```
```suggestion
make iso BASE="${MYBASE}" RELEASE="${MYRELEASE}" ARCH="${MYARCH}" ROOTPW_HASH="*"
```

(assuming UPLOAD is NO by default as suggested above)

grembo commented on build.sh:
```diff
@@ -164,18 +195,24 @@ if [ -f "${OUTISO}" ]; then
 # scp to distribution site
-if [ "${UPLOAD}" -ne 0 ]; then
-    scp -P "${remoteport}" "${OUTISO}" "${remoteuser}"@"${remotehost}":"${remotepath}"/"${OUTISO}"
+if [ "$UPLOAD" = "YES" ] && [ -n "${CFG_SSH_REMOTEHOST}" ]; then
+    scp "${OUTISO}" "${CFG_SSH_REMOTEHOST}":"${CFG_SSH_REMOTEPATH}"
+else
+    exit_error "CFG_SSH_REMOTEHOST is unset"
 fi
```
This logic is flawed, the correct logic would be:

```sh
if [ "$UPLOAD" = "YES" ]; then
    if [ -z "${CFG_SSH_REMOTEHOST}" ]; then
        exit_error "CFG_SSH_REMOTEHOST is unset"
    fi
    scp "${OUTISO}" "${CFG_SSH_REMOTEHOST}":"${CFG_SSH_REMOTEPATH}"
fi
```

But I would suggest to move all of this higher up to when settings.sh is read, so we fail early before the image was built etc, therefore this could can change to:

 ```suggestion
if [ "$UPLOAD" = "YES" ]; then
    scp "${OUTISO}" "${CFG_SSH_REMOTEHOST}":"${CFG_SSH_REMOTEPATH}"
fi
```

grembo commented on settings.sh.sample:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
```
This line can go away, no shebang needed when sourcing.

In case this file's executable bit is set, please change file mode to 0644 (didn't check myself)

grembo commented on build.sh:
```diff
@@ -5,80 +5,109 @@
 # 2022-07-31: adding extra rc.local, setting rc build script
```
In case build.sh is **not** executable, please change to 0755 (didn't check)

bretton replied:
this is fixed. I had it set to non-executable because I'm copying to build-rc.sh and running that with some changes to image name. This is non-issue now as default is build without upload.

grembo commented on build.sh:
```diff
@@ -5,80 +5,109 @@
+	-b Build without uploading
+	-u Build with upload to remote host
+	-k /path/to/authorized_keys (can safely ignore, another opportunity to copy in SSH keys on image boot!)
+	
```
I think this tab character can go away (so no trailing whitespace) and it should still format correctly (probably best to test once). (the one in line 38, the empty line)

grembo commented on build.sh:
```diff
@@ -5,80 +5,109 @@
 # check remote settings
-if [ -f "${MYREMOTECONFIG}" ]; then
-    source "${MYREMOTECONFIG}"
+# shellcheck source=/dev/null
```
## General

I would suggest to move the shellcheck comment directly above the source command, so it's absolutely clear what it refers to.

## My first suggestion (read only if curious)

I would also just read the parameters in when UPLOAD is YES (right now, that's the only time when it's needed, even though this might change in the future) and validate parameters.

Finally, config variables should be set to empty to make sure they're set using settings.sh (otherwise there could be an empty settings file and variables come from the environment, which could be confusing):

```suggestion
if [ "$UPLOAD" = "YES" ]; then
    CFG_SSH_REMOTEHOST=
    CFG_SSH_REMOTEPATH=
    if [ -f "${BASEDIR}/settings.sh" ]; then
        # shellcheck source=/dev/null
        source "${BASEDIR}/settings.sh"
    else
        exit_error "Please copy settings.sh.sample to settings.sh and set parameters"
    fi
    if [ -z "${CFG_SSH_REMOTEHOST}" ]; then
        exit_error "CFG_SSH_REMOTEHOST not configured in settings.sh"
    fi
fi
```

## My preferred solution

The alternative would be to not to wipe CFG_SSH_... at all, but in this case I would suggest to make settings.sh completely optional (as users could pass the variables in over the command line without creating settings.sh).

If you want to go that route, the suggestion would be:

```suggestion
if [ -f "${BASEDIR}/settings.sh" ]; then
    # shellcheck source=/dev/null
    source "${BASEDIR}/settings.sh"
fi
if [ "$UPLOAD" = "YES" ] && [ -z "${CFG_SSH_REMOTEHOST}" ]; then
    exit_error "CFG_SSH_REMOTEHOST not configured in settings.sh"
fi
```

To be honest, I like this second option better, as it allows users to do:

    CFG_SSH_REMOTEHOST=myhost ./build.sh -u

grembo replied:
I just had another idea: What about providing a default for CFG_SSH_REMOTEHOST? (e.g., `depenguin-me`).

Then all users would need to do is add an entry to their `~/.ssh/config` like this one:

```
Host depenguin-me
   User myuser
   HostName myactualhost.example.org
   IdentityFile ~/.ssh/byprivatekey
```

And they could run `./build.sh -u`

In this case, the code would be:

```suggestion
if [ -f "${BASEDIR}/settings.sh" ]; then
    # shellcheck source=/dev/null
    source "${BASEDIR}/settings.sh"
fi
if [ -z "${CFG_SSH_REMOTEHOST}" ]; then
    CFG_SSH_REMOTEHOST=depenguin-me
fi
```

**Note:** Suggestions don't really work when there are deleted lines, so you can't simply apply it in github's web u

bretton replied:
adopting this approach. 

tests:
- build image works
- run image with flags for keys works (login, sudo to root)

bretton replied:
thanks for the feedback!

grembo reviewed (changes requested):
@bretton I went quite a bit into detail here, I hope it's worthwhile

grembo commented on build.sh:
```diff
@@ -5,80 +5,106 @@
 else
-    echo "Please copy settings.cfg.sample to settings.cfg and set remote parameters"
-    exit
+    exit_error "Please copy settings.sh.sample to settings.sh and set parameters"
+fi
```
These two lines have to go, so it's possible to pass in `CFG_SSH_REMOTEHOST` / have a default value when no settings.sh was created.

grembo replied:
I marked the wrong lines, I meant these two:

```
else
    exit_error "Please copy settings.sh.sample to settings.sh and set parameters"
```

bretton replied:
fixed

grembo reviewed (changes requested):
Almost there :)

bretton commented:
shellcheck exclusion was added to get around an error that cropped up. now not appearing any more. 

tests:
- build image works
- run image with flags for keys works (login, sudo to root)

grembo reviewed (approved):
Ready to land ;)

bretton commented:
live build test success.
live run.sh test success, login, sudo to root

```
SHA256 (mfsbsd-13.1-RELEASE-amd64.iso) = 059adc4b9352300c2b5c3c07f438471b42251c13bbf902df8885c94264883f8d
```

bretton committed:
Adjustments based on PR review feedback

Earlier logic was checking for arguments, removed. Default is to build now.

To upload too, set the `-u` flag.

Adjusted settings.sh approach too

bretton committed:
Making build.sh executable

Tweaking settings.sh.sample even though not needed specifically any more.

bretton committed:
Fixing minor fixes missed earlier

bretton committed:
Updating README to remove setting executable

bretton committed:
Removing items based on PR feedback

Removed line with error about missing settings.sh

Removed shellcheck exclusion for iso mount
